services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala

```diff
@@ -488,6 +567,94 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
+                                                 ),
+                                                 uniqueActorName(subId)
+        )
+        subMetadataBuilder ! GetCost(WorkflowId.fromString(subId))
+      }
+      goto(WaitingForSubWorkflowCost) using HasReceivedEventsData(target,
+                                                                  originalRequest,
+                                                                  metadataResponse.query,
+                                                                  metadataResponse.eventList,
+                                                                  Map.empty,
+                                                                  subWorkflowIds.size
+      )
+    }
+  }
+
+  def buildCostAndStop(id: WorkflowId,
+                       status: String,
+                       eventsList: Seq[MetadataEvent],
+                       expandedValues: Map[String, JsValue],
+                       target: ActorRef,
+                       originalRequest: BuildMetadataJsonAction
+  ): State = {
+
+    val metadataEvents = MetadataBuilderActor.parse(groupEvents(eventsList), expandedValues)
+
+    // Walk the structured metadata to attempt to compute cost for each call/shard/attempt in this workflow
+    val callCostsWithErrors: List[ErrorOr[BigDecimal]] = for {
+      wfObj <- extractFromJsAs[JsObject](metadataEvents, id.toString).toList
+      callsObj <- extractFromJsAs[JsObject](wfObj, "calls").toList
+      singleCallName <- callsObj.fields.keys.toList
+      singleCallAttemptList <- extractFromJsAs[JsArray](callsObj, singleCallName).toList
+      singleCallAttempt <- singleCallAttemptList.elements.toList
+      singleCallAttemptCost = computeCost(singleCallName, singleCallAttempt)
+    } yield singleCallAttemptCost
+
+    val callCost: BigDecimal = callCostsWithErrors.collect { case (Valid(b)) => b }.sum
+    val costErrors: Vector[JsString] =
+      callCostsWithErrors
+        .collect { case (Invalid(e)) => e.toList.mkString(", ") }
+        .map(JsString(_))
+        .toVector
+
+    // Get the cost and error collection for any subworkflows that are part of this workflow
+    val subworkflowCost = expandedValues.values.map {
+      case o: JsObject =>
+        extractFromJsAs[JsNumber](o, "cost").map(_.value).getOrElse(BigDecimal(0))
+      case _ => BigDecimal(0)
+    }.sum
+
+    val subworkflowErrors = expandedValues.values.flatMap {
+      case o: JsObject =>
+        extractFromJsAs[JsArray](o, "errors").getOrElse(JsArray.empty).elements
+      case _ => JsArray.empty.elements
+    }
+
+    val resp = JsObject(
+      Map(
+        WorkflowMetadataKeys.Id -> JsString(id.toString),
+        WorkflowMetadataKeys.Status -> JsString(status),
+        "currency" -> JsString(DefaultCurrency.getCurrencyCode),
+        "cost" -> JsNumber(callCost + subworkflowCost),
+        "errors" -> JsArray(costErrors ++ subworkflowErrors)
```

Comment:
If there are any errors when computing cost, should the `cost` field be 0? Or is the intention for it to always reflect the approximate cost of jobs that were successfully calculated?

Reply:
The latter - then the caller can decide how much they want to believe that cost. My thinking is that if it's a huge workflow and we failed to compute the cost for just one task, it's still useful to know the cost of the rest of it.